schemas.py

```diff
@@ -1,14 +1,144 @@
-from marshmallow import Schema, fields
+from marshmallow_jsonapi import Schema, fields
 
 
 class BaseSchema(Schema):
+    """
+    Base class for all Schemas.
+    Extends marshmallow.Schema and Implements utility functions to validate
+    inputs for the classes, to get schemas and serialize models data.
+    """
     @classmethod
-    def json(cls, obj):
-        return cls().dump(obj)
+    def json(cls, obj, include_data=[]):
+        """
+        http://marshmallow.readthedocs.io/en/latest/quickstart.html#serializing-objects-dumping
+        Serialize obj by passing it to schema's dump method, which returns
+        the formatted result.
+
+        :param obj object: The object to serialize
+        :param include_data list: a list of fields inside the object to include
+                                  inside the serialized response
+        :returns: (data, errors)
+        """
+
+        serialized = cls(include_data=include_data).dump(obj)
+        return serialized.data, serialized.errors
+
+    @classmethod
+    def validate_input(cls, jsondata):
+        """"
+        http://marshmallow.readthedocs.io/en/latest/quickstart.html#schema-validate
+        validate an input json data against the schema of its relative object
+        schema
+        :returns:
+        * True if validation is ok
+        * Errors dict {<key>: <error>} if validation does not pass
+        """
+
+        errors = cls().validate(jsondata)
+        if not errors:
+            return True, {}
+        return False, errors
+
+
+class ItemSchema(BaseSchema):
+    """
+    Schema for models.Item.
+    """
+
+    class Meta:
+        type_ = 'item'
+        self_url = '/items/{id}'
+        self_url_kwargs = {'id': '<id>'}
+        self_url_many = '/items/'
+
+    id = fields.Str(dump_only=True, attribute='item_id')
+    name = fields.Str()
+    price = fields.Float()
+    description = fields.Str()
+
+
+class OrderSchema(BaseSchema):
+    """
+    Schema for models.Order.
+    Include an include relationship with the items included in the order and
+    the user that created it.
+    """
+
+    class Meta:
+        type_ = 'order'
+        self_url_many = '/orders/'
+
+    id = fields.Str(dump_only=True, attribute='order_id')
+    date = fields.DateTime(attribute='created_at')
+    total_price = fields.Float()
+    delivery_address = fields.Str()
+
+    # Uses the OrderItem table and OrderItemSchema to serialize a json object
+    # representing each item in the order
+    items = fields.Relationship(
+        many=True, include_resource_linkage=True,
+        type_='item', schema='OrderItemSchema',
+        dump_only=True, id_field='item.item_id',
+        attribute='order_items',
+    )
+
+    user = fields.Relationship(
+        include_resource_linkage=True,
+        type_='user', schema='UserSchema',
+        dump_only=True, id_field='user_id',
+    )
+
+    @classmethod
+    def json(cls, obj, include_data=['items', 'user']):
```

Comment:
Sì, il valore ritornato da `json` è una struttura jsonapi.

Il metodo esegue una `cls().dump(<obj>)`, che ritorna un oggetto marshmallow con dentro `data` - l'oggetto serializzato - ed `errors` - dict con dentro eventuali errori di parsing. le due `str` passate ad `include_data` fanno riferimento al nome degli attributi della classe chiamante (nel caso in oggetto `OrderSchema`), che devono essere istanze di `fields.Relationship` con un attr `schema=<NomeSchema>` (come pointer alla classe o come stringa).

I default li ho messi partendo dal presupposto che quando serializziamo un ordine vogliamo includere anche l'user che ha creato l'ordine e gli items dell'ordine, nel caso di `OrderSchema`